### PR TITLE
sql: Implement GRANT/REVOKE for all objects

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds the syntax to GRANT and REVOKE that allows users to
grant or revoke privileges on all objects of a certain type within a
schema.

### Motivation
This PR adds a feature that has not yet been specified.

### Tips for reviewer
  * ~~This builds off of https://github.com/MaterializeInc/materialize/pull/19590, all new changes are in the final commit.~~
  * ~~This is blocked by https://github.com/MaterializeInc/materialize/pull/19535 for testing.~~

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add the syntax to GRANT and REVOKE that allows users to grant or revoke privileges on all objects of a certain type within a schema.

### CODE REVIEW

File: src/sql-parser/src/parser.rs [hunk 1]
```diff
@@ -6032,115 +6032,179 @@ impl<'a> Parser<'a> {
...
+    }
+
+    /// Bail out if the current token is not an object type suitable for a GRANT/REVOKE, or consume
+    /// and return it if it is.
+    fn expect_grant_revoke_object_type(
+        &mut self,
+        statement_type: &str,
+    ) -> Result<ObjectType, ParserError> {
+        // If the object type is omitted, then it is assumed to be a table.
+        let object_type = self.parse_object_type().unwrap_or(ObjectType::Table);
```
Comment (mjibson) [inline src/sql-parser/src/parser.rs:6146]: Is this copying postgres? Could we instead make it not assumed but required?

File: src/sql-parser/src/parser.rs [hunk 2]
```diff
@@ -6032,115 +6032,179 @@ impl<'a> Parser<'a> {
...
+    }
+
+    /// Bail out if the current token is not a plural object type suitable for a GRANT/REVOKE, or consume
+    /// and return it if it is.
+    fn expect_grant_revoke_plural_object_type_in_schema(
+        &mut self,
+        statement_type: &str,
+    ) -> Result<ObjectType, ParserError> {
+        let object_type = self.expect_plural_object_type().map_err(|_| {
+            // Limit the error message to allowed object types.
```
Comment (mjibson) [inline src/sql-parser/src/parser.rs:6157]: I don't understand this. Why are the 4 below the only allowed object types?

**Review by mjibson — APPROVED** [review]

File: src/sql-parser/src/parser.rs (hunk 1 above)
Comment (jkosh44) [reply src/sql-parser/src/parser.rs:6146]: This does copy postgres. I'm a bit torn on this. If the parsed object type always matched the actual object type then I would prefer it to be required. However, when granting privileges on views, materialized views, and sources, you have to specify table. So for example let's say there's some view v, then `GRANT _ ON VIEW v TO _` is invalid, you must say `GRANT _ ON TABLE v TO _` or `GRANT _ ON v TO _`. So it is nice that you can omit table when the object isn't actually a table.

We could also allow `GRANT _ ON VIEW v TO _`, but we'd probably also want the other two version to be valid anyway for postgres compatibility.

File: src/sql-parser/src/parser.rs (hunk 2 above)
Comment (jkosh44) [reply src/sql-parser/src/parser.rs:6157]: Those are the only valid object types that exist in a schema. All relations are covered by `TABLES`.

File: src/sql-parser/src/parser.rs (hunk 1 above)
Comment (mjibson) [reply src/sql-parser/src/parser.rs:6146]: Oh I forgot about the silly tables are views rule. Ok, fine is as.

File: src/sql-parser/src/parser.rs (hunk 2 above)
Comment (mjibson) [reply src/sql-parser/src/parser.rs:6157]: The table rule is so weird.

Comment (jkosh44) [reply src/sql-parser/src/parser.rs:6157]: Agreed, it's completely inconsistent with every other command.

**Pull request merged**